File: src/storage/object_store.cpp
#include "flux/storage/object_store.hpp"
#include "flux/util/filesystem.hpp"
#include "flux/util/compression.hpp"
#include <fmt/format.h>
#include <fstream>

namespace flux {

ObjectStore::ObjectStore(Path objects_dir) 
    : objects_dir_(std::move(objects_dir))
    , pack_manager_(objects_dir_ / "pack") {
    // Attempt to load existing packs, ignore errors in constructor
    (void)pack_manager_.load_packs();
}

Result<ObjectId> ObjectStore::write(ObjectType type, std::span<const uint8_t> data, HashAlgorithm algo) {
    // Compute object ID
    ObjectId id = ObjectId::compute(algo, data, type);
    
    // Check if already exists
    if (exists(id)) {
        return id;
    }
    
    // Write as loose object
    auto result = write_loose(id, type, data);
    if (!result) {
        return flux::unexpected(result.error());
    }
    
    return id;
}

Result<Bytes> ObjectStore::read(const ObjectId& id) {
    // Try loose objects first
    auto loose_result = read_loose(id);
    if (loose_result) {
        return loose_result;
    }
    
    // Try pack files
    if (pack_manager_.contains(id)) {
        return pack_manager_.read_object(id);
    }
    
    return flux::unexpected(fmt::format("Object not found: {}", id.to_hex()));
}

bool ObjectStore::exists(const ObjectId& id) {
    return Filesystem::exists(get_loose_path(id));
}

Result<ObjectType> ObjectStore::get_type(const ObjectId& id) {
    Path obj_path = get_loose_path(id);
    std::ifstream file(obj_path, std::ios::binary);
    if (!file) {
        return flux::unexpected(fmt::format("Object not found: {}", id.to_hex()));
    }
    
    char type_byte;
    if (!file.get(type_byte)) {
        return flux::unexpected("Failed to read type byte");
    }
    
    return static_cast<ObjectType>(static_cast<uint8_t>(type_byte));
}

Result<void> ObjectStore::remove(const ObjectId& id) {
    return Filesystem::remove(get_loose_path(id));
}

Path ObjectStore::get_loose_path(const ObjectId& id) const {
    // Use first 2 hex chars as subdirectory for better filesystem performance
    std::string hex = id.to_hex();
    
    // Extract algorithm prefix and hash
    auto colon_pos = hex.find(':');
    std::string algo_prefix = hex.substr(0, colon_pos);
    std::string hash_hex = (colon_pos != std::string::npos) ? hex.substr(colon_pos + 1) : hex;
    
    if (hash_hex.length() < 2) {
        return objects_dir_ / "loose" / "invalid";
    }
    
    // Create path: objects/loose/algo/ab/cdef...
    return objects_dir_ / "loose" / algo_prefix / hash_hex.substr(0, 2) / hash_hex.substr(2);
}

Result<void> ObjectStore::write_loose(const ObjectId& id, ObjectType type, std::span<const uint8_t> data) {
    Path obj_path = get_loose_path(id);
    
    // Create parent directories
    auto dir_result = Filesystem::create_directories(obj_path.parent_path());
    if (!dir_result) {
        return dir_result;
    }
    
    // Format: <type_byte><compressed_data>
    Bytes obj_data;
    obj_data.push_back(static_cast<uint8_t>(type));
    
    // Compress data
    auto compressed_result = Compression::compress(data);
    if (!compressed_result) {
        return flux::unexpected(compressed_result.error());
    }
    
    obj_data.insert(obj_data.end(), compressed_result->begin(), compressed_result->end());
    
    // Write atomically
    return Filesystem::write_file(obj_path, obj_data);
}

Result<Bytes> ObjectStore::read_loose(const ObjectId& id) {
    Path obj_path = get_loose_path(id);
    
    auto data_result = Filesystem::read_file(obj_path);
    if (!data_result) {
        return data_result;
    }
    
    if (data_result->empty()) {
        return flux::unexpected("Invalid object: empty file");
    }
    
    // Skip type byte
    std::span<const uint8_t> compressed_data(data_result->data() + 1, data_result->size() - 1);
    
    // Decompress
    return Compression::decompress(compressed_data);
}

} // namespace flux

File: src/storage/object_store.zig
const std = @import("std");
const types = @import("../core/types.zig");
const object_id_module = @import("../core/object_id.zig");
const filesystem = @import("../util/filesystem.zig");
const compression = @import("../util/compression.zig");

const ObjectId = object_id_module.ObjectId;
const ObjectType = types.ObjectType;
const HashAlgorithm = types.HashAlgorithm;
const Filesystem = filesystem.Filesystem;
const Compression = compression.Compression;

/// ObjectStore - Content-addressed object storage
pub const ObjectStore = struct {
    objects_dir: []const u8,
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator, objects_dir: []const u8) ObjectStore {
        return .{
            .allocator = allocator,
            .objects_dir = objects_dir,
        };
    }

    /// Write object to store
    pub fn write(self: *ObjectStore, obj_type: ObjectType, data: []const u8, algo: HashAlgorithm) !ObjectId {
        // Compute object ID
        const id = try ObjectId.compute(self.allocator, algo, data, obj_type);
        errdefer id.deinit();

        // Check if already exists
        if (try self.exists(&id)) {
            return id;
        }

        // Write as loose object
        try self.writeLoose(&id, obj_type, data);
        return id;
    }

    /// Read object from store
    pub fn read(self: *ObjectStore, id: *const ObjectId) ![]u8 {
        // Try loose objects
        return self.readLoose(id) catch |err| {
            if (err == error.FileNotFound) {
                // Could try pack files here
                return error.ObjectNotFound;
            }
            return err;
        };
    }

    /// Check if object exists
    pub fn exists(self: *ObjectStore, id: *const ObjectId) !bool {
        const path = try self.getLoosePath(id);
        defer self.allocator.free(path);
        return Filesystem.exists(path);
    }

    /// Get object type
    pub fn getType(self: *ObjectStore, id: *const ObjectId) !ObjectType {
        const path = try self.getLoosePath(id);
        defer self.allocator.free(path);

        const data = try Filesystem.readFile(self.allocator, path);
        defer self.allocator.free(data);

        if (data.len == 0) return error.InvalidObject;

        return @enumFromInt(data[0]);
    }

    /// Remove object
    pub fn remove(self: *ObjectStore, id: *const ObjectId) !void {
        const path = try self.getLoosePath(id);
        defer self.allocator.free(path);
        try Filesystem.remove(path);
    }

    /// Get loose object path
    fn getLoosePath(self: *ObjectStore, id: *const ObjectId) ![]u8 {
        const hex = try id.toHex(self.allocator);
        defer self.allocator.free(hex);

        // Extract algorithm and hash
        const colon_pos = std.mem.indexOf(u8, hex, ":") orelse hex.len;
        const algo_prefix = hex[0..colon_pos];
        const hash_hex = if (colon_pos < hex.len) hex[colon_pos + 1 ..] else hex;

        if (hash_hex.len < 2) {
            return try std.fs.path.join(self.allocator, &[_][]const u8{ self.objects_dir, "loose", "invalid" });
        }

        // Create path: objects/loose/algo/ab/cdef...
        const parts = [_][]const u8{
            self.objects_dir,
            "loose",
            algo_prefix,
            hash_hex[0..2],
            hash_hex[2..],
        };

        return try std.fs.path.join(self.allocator, &parts);
    }

    /// Write loose object
    fn writeLoose(self: *ObjectStore, id: *const ObjectId, obj_type: ObjectType, data: []const u8) !void {
        const obj_path = try self.getLoosePath(id);
        defer self.allocator.free(obj_path);

        // Format: <type_byte><compressed_data>
        var obj_data = std.ArrayList(u8).init(self.allocator);
        defer obj_data.deinit();

        try obj_data.append(@intFromEnum(obj_type));

        // Compress data
        const compressed = try Compression.compress(self.allocator, data, 3);
        defer self.allocator.free(compressed);

        try obj_data.appendSlice(compressed);

        // Write atomically
        try Filesystem.writeFile(obj_path, obj_data.items);
    }

    /// Read loose object
    fn readLoose(self: *ObjectStore, id: *const ObjectId) ![]u8 {
        const obj_path = try self.getLoosePath(id);
        defer self.allocator.free(obj_path);

        const data = try Filesystem.readFile(self.allocator, obj_path);
        errdefer self.allocator.free(data);

        if (data.len == 0) return error.InvalidObject;

        // Skip type byte and decompress
        const compressed_data = data[1..];

        const decompressed = try Compression.decompress(self.allocator, compressed_data);
        self.allocator.free(data);

        return decompressed;
    }
};
